plugins/woocommerce/client/admin/client/wp-admin-scripts/email-editor-integration/blocks/coupon-code/edit.tsx

```diff
@@ -0,0 +1,352 @@
+ *
+ * @param {BlockEditProps} props - Block properties.
+ * @return {JSX.Element} The edit component.
+ */
+export function Edit( props: BlockEditProps ): JSX.Element {
+	const { attributes, setAttributes } = props;
+	const couponCode = attributes.couponCode as string;
+
+	const {
+		className: blockClassName = '',
+		style: blockStyle,
+		...wrapperProps
+	} = useBlockProps();
+	const [ searchValue, setSearchValue ] = useState( '' );
+	const [ coupons, setCoupons ] = useState< Coupon[] >( [] );
+	const [ isLoading, setIsLoading ] = useState( true );
+	const [ isTruncated, setIsTruncated ] = useState( false );
+
+	// Handler for creating a new coupon - uses a filter so integrators can customize behavior
+	const handleCreateCoupon = () => {
+		const createCouponHandler = applyFilters(
+			'woocommerce_email_editor_create_coupon_handler',
+			() => {
+				// Default fallback: open in new tab
+				window.open(
+					'/wp-admin/post-new.php?post_type=shop_coupon',
+					'_blank'
+				);
+			}
+		);
+
+		if ( typeof createCouponHandler === 'function' ) {
+			createCouponHandler();
+		}
+	};
+
+	// Fetch coupons from WooCommerce API with pagination
+	useEffect( () => {
+		const controller = new AbortController();
+		const PER_PAGE = 100;
+		const MAX_PAGES = 10; // Safety cap to prevent infinite loops
+
+		const fetchCoupons = async () => {
+			try {
+				setIsLoading( true );
+				setIsTruncated( false );
+
+				const allCoupons: Coupon[] = [];
+				let currentPage = 1;
+				let totalPages: number | null = null;
+				let hasMorePages = true;
+
+				while ( hasMorePages && currentPage <= MAX_PAGES ) {
+					const response = await apiFetch< Response >( {
+						path: `/wc/v3/coupons?per_page=${ PER_PAGE }&page=${ currentPage }`,
+						signal: controller.signal,
+						parse: false,
+					} );
+
+					// Extract pagination headers
+					const totalPagesHeader =
+						response.headers.get( 'X-WP-TotalPages' );
+
+					if ( totalPagesHeader ) {
+						totalPages = parseInt( totalPagesHeader, 10 );
+					}
+
+					// Parse response body
+					const pageCoupons: Coupon[] = await response.json();
+
+					allCoupons.push( ...pageCoupons );
+
+					// Determine if we should continue fetching
+					if ( totalPages !== null ) {
+						// Use header-based pagination if available
+						hasMorePages = currentPage < totalPages;
+					} else {
+						// Fallback: check if we got fewer items than per_page
+						hasMorePages = pageCoupons.length === PER_PAGE;
+					}
+
+					// Safety check: if we've reached max pages and there are more, mark as truncated
+					if ( currentPage >= MAX_PAGES && hasMorePages ) {
+						setIsTruncated( true );
+						break;
+					}
+
+					currentPage++;
+				}
+
+				setCoupons( allCoupons );
+			} catch ( error ) {
+				if ( error instanceof Error && error.name === 'AbortError' ) {
+					return;
+				}
+				// eslint-disable-next-line no-console
+				console.error( 'Error fetching coupons:', error );
+				setCoupons( [] );
```

Comment:
Please correct me if I'm wrong, but there's no need to set coupons to an empty array again since we initially set it as an empty array.